torchtune/modules/model_fusion/_fusion.py

```diff
@@ -463,18 +500,294 @@ def forward(
             - d_e: encoder embed dim
             - m_s: max seq len
         """
+        if encoder_input is not None and encoder_input.keys() != self.encoders.keys():
+            raise ValueError(
+                f"Found mismatched keys in encoder_input and instantiated encoders. "
+                f"Got {encoder_input.keys()}, expected {self.encoders.keys()}."
+            )
         # During decoding, encoder_input will only be provided
         # for new inputs. Previous encoder outputs are cached
         # in the decoder cache.
         encoder_embed = None
         if encoder_input is not None:
-            encoder_embed = self.encoder(**encoder_input)
+            encoder_embed = {
+                encoder: self.encoders[encoder](**encoder_input[encoder])
+                for encoder in encoder_input
+            }
+
+        # Currently, only a single encoder is supported, so we need
+        # to get the encoder key manually. When multiple encoders are
+        # supported, this can be removed.
+        decoder_encoder_input = (
+            list(encoder_embed.values())[0] if encoder_embed is not None else None
+        )
 
         output = self.decoder(
             tokens=tokens,
             mask=mask,
-            encoder_input=encoder_embed,
+            encoder_input=decoder_encoder_input,
             encoder_mask=encoder_mask,
             input_pos=input_pos,
         )
         return output
+
+
+class EarlyFusionModel(nn.Module):
```

Comment:
I don't have an incredibly strong opinion here but it might be clearest to put this in a separate file (e.g. deep_fusion.py and early_fusion.py)

Reply:
yeah, could split to deep_fusion, early_fusion, and fusion_layers.py